### ISSUE TITLE
S3 async downloads complete but file 0 bytes

### ISSUE THREAD

**Issue by ben-manes** [aws/aws-sdk-java-v2#721]
After upgrading to the new sdk, I am observing the download future completes prior to the file appearing. In the following code, the assertion fails sporadically. The file does exist on S3, so it appears to be a race condition in the client.

```java
public Path download(Path directory, String key) {
  var request = GetObjectRequest.builder().bucket(bucket).key(normalizeKey(key)).build();
  Stopwatch stopwatch = Stopwatch.createStarted();
  try {
    Path destination = directory.resolve(request.key());
    Files.createDirectories(destination.getParent());
    s3.getObject(request, toFile(destination)).join();
    checkState(Files.exists(destination) && Files.size(destination) > 0L,
        "Failed to download from S3: %s", request.key());

    logger.info("Completed S3 download of {} in {}", request.key(), stopwatch);
    return destination;
  } catch (Exception e) {
    throw new RuntimeException("Failed to download from S3: " + request.key(), e);
  }
}
```

```
java.lang.IllegalStateException: Failed to download from S3: c8dbae15-807a-48e1-8a4a-4fb4418bee73/component/image/1537382453460/7521bcaa-dc86-49f3-b248-989e21851cfc/page2.txt
	at com.google.common.base.Preconditions.checkState(Preconditions.java:588)
```

**Comment by ben-manes** [comment]
I also see a lot of these errors. I think this became more unstable in `preview-12`...

```
software.amazon.awssdk.http.SdkCancellationException: Subscriber cancelled before all events were published
	at software.amazon.awssdk.http.nio.netty.internal.ResponseHandler$PublisherAdapter$1.onCancel(ResponseHandler.java:214)
	at software.amazon.awssdk.http.nio.netty.internal.ResponseHandler$OnCancelSubscription.cancel(ResponseHandler.java:279)
	at software.amazon.awssdk.core.internal.async.FileAsyncResponseTransformer$FileSubscriber$1.failed(FileAsyncResponseTransformer.java:137)
	at software.amazon.awssdk.core.internal.async.FileAsyncResponseTransformer$FileSubscriber$1.failed(FileAsyncResponseTransformer.java:119)
	at java.base/sun.nio.ch.Invoker.invokeUnchecked(Invoker.java:129)
	at java.base/sun.nio.ch.SimpleAsynchronousFileChannelImpl$3.run(SimpleAsynchronousFileChannelImpl.java:389)
	... 3 common frames omitted
Wrapped by: software.amazon.awssdk.core.exception.SdkClientException: null
	at software.amazon.awssdk.core.exception.SdkClientException$BuilderImpl.build(SdkClientException.java:97)
	at software.amazon.awssdk.core.internal.http.pipeline.stages.AsyncRetryableStage$RetryExecutor.handle(AsyncRetryableStage.java:160)
	... 9 common frames omitted
Wrapped by: java.util.concurrent.CompletionException: software.amazon.awssdk.core.exception.SdkClientException
	at java.base/java.util.concurrent.CompletableFuture.encodeThrowable(CompletableFuture.java:331)
	at java.base/java.util.concurrent.CompletableFuture.completeThrowable(CompletableFuture.java:346)
	at java.base/java.util.concurrent.CompletableFuture$UniApply.tryFire(CompletableFuture.java:632)
	at java.base/java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:506)
	at java.base/java.util.concurrent.CompletableFuture.completeExceptionally(CompletableFuture.java:2088)
	at software.amazon.awssdk.core.internal.http.pipeline.stages.AsyncRetryableStage$RetryExecutor.handle(AsyncRetryableStage.java:172)
	at software.amazon.awssdk.core.internal.http.pipeline.stages.AsyncRetryableStage$RetryExecutor.lambda$execute$0(AsyncRetryableStage.java:138)
	at java.base/java.util.concurrent.CompletableFuture.uniHandle(CompletableFuture.java:930)
	at java.base/java.util.concurrent.CompletableFuture$UniHandle.tryFire(CompletableFuture.java:907)
	at java.base/java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:506)
	at java.base/java.util.concurrent.CompletableFuture.completeExceptionally(CompletableFuture.java:2088)
	at software.amazon.awssdk.core.internal.http.pipeline.stages.MakeAsyncHttpRequestStage$Completable.lambda$completeExceptionally$1(MakeAsyncHttpRequestStage.java:239)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1135)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
	at java.base/java.lang.Thread.run(Thread.java:844)
```

**Comment by dagnir** [comment]
@ben-manes Can you provide additional details about your environment (OS and JRE version?)

**Comment by ben-manes** [comment]
The OS is old (Ubuntu 14.04.5 LTS) on EC2, so we should update it. The JDK is new at 10.0.2. The configuration is pretty benign to provide our own `S3` helper class that all of our usages go through (e.g. provides above method).

```java
public final class AwsModule extends AbstractModule {
  private final Config config = ConfigFactory.load().getConfig(getClass().getPackage().getName());

  @Override
  protected void configure() {
    install(new S3Module());
    Multibinder.newSetBinder(binder(), SdkClient.class);
    Multibinder.newSetBinder(binder(), Service.class).addBinding().to(SdkClientService.class);
  }

  @Provides
  Region providesRegion() {
    return Region.of(config.getString("region"));
  }

  @Provides
  AwsCredentialsProvider providesCredentials() {
    return StaticCredentialsProvider.create(
        AwsBasicCredentials.create(config.getString("access-key"), config.getString("secret-key")));
  }

  @Override
  public boolean equals(Object o) {
    return (o instanceof AwsModule);
  }

  @Override
  public int hashCode() {
    return getClass().hashCode();
  }

  @Singleton
  private static final class SdkClientService extends AbstractIdleService {
    private final Set<SdkClient> clients;

    @Inject
    SdkClientService(Set<SdkClient> clients) {
      this.clients = clients;
    }

    @Override
    protected void startUp() {}

    @Override protected void shutDown() {
      for (var client : clients) {
        client.close();
      }
    }
  }
}
```

```java
public final class S3Module extends AbstractModule {
  private final Config config = ConfigFactory.load().getConfig(getClass().getPackage().getName());

  @Override
  protected void configure() {
    var clients = Multibinder.newSetBinder(binder(), SdkClient.class);
    clients.addBinding().to(CloudFrontAsyncClient.class);
    clients.addBinding().to(S3AsyncClient.class);
  }

  @Provides @Singleton
  S3AsyncClient providesS3AsyncClient(AwsCredentialsProvider credentials) {
    return S3AsyncClient.builder()
        .region(Region.of(config.getString("region")))
        .credentialsProvider(credentials)
        .build();
  }

  @Provides @Singleton
  CloudFrontAsyncClient providesCloudFrontClient(AwsCredentialsProvider credentials) {
    return CloudFrontAsyncClient.builder()
        .credentialsProvider(credentials)
        .region(AWS_GLOBAL)
        .build();
  }

  @Provides @Singleton
  S3 providesS3(S3AsyncClient s3, CloudFrontAsyncClient cloudFront, Tika tika)
      throws URISyntaxException {
    URI cdn = config.hasPath("cdn") ? new URI(config.getString("cdn")) : null;
    String distributionId = (cdn == null) ? null : config.getString("distribution-id");
    URI imageCdn = config.hasPath("image-cdn") ? new URI(config.getString("image-cdn")) : null;
    return new S3(tika, s3, config.getString("bucket"), cloudFront,
        cdn, imageCdn, distributionId, config.getDuration("max-age"));
  }
}
```

**Comment by ben-manes** [comment]
For now this v11 is non-blocking since I can fallback to an httpClient when the file doesn't exist. We already have that code if the URL is not to our non-image CDN (as images are resized dynamically by the URI). I'd probably get a lot of `SdkCancellationException` spam if upgrading to v12, so holding off on that.

**Comment by ben-manes** [comment]
I am also seeing cases where an asynchronous PUT does not complete. This might be a related race in your event handling. I have code that optimistically uploads during a workflow process, keeping the files cached locally for downstream steps if queried for. Prior to completing, the workflow blocks on the accumulated uploads until all complete before finishing successfully. Unfortunately, some futures never complete and it hangs the workflow if a timeout isn't set. I think it could all be interrelated, so adding here rather than as a new ticket. This is relatively rare, but daily at volume (many workflows with only a few files).

```java
// S3 utility class:

/**
 * Creates a file on S3 with the given key as the file name. The request may fail if the key
 * exists and the bucket does not allow overwrites (it is advisable to include a UUID as part of
 * the key). Note that the URI may be optimistically obtained using {@link #getUri(String)}.
 *
 * @param key the name of the file in the bucket
 * @param path the path to the file
 * @return the location to the file
 */
public CompletableFuture<URI> upload(String key, Path path) {
  try {
    var stopwatch = Stopwatch.createStarted();
    var request = PutObjectRequest.builder()
        .contentType(tika.detect(path))
        .cacheControl(cacheControl)
        .key(normalizeKey(key))
        .acl(PUBLIC_READ)
        .bucket(bucket)
        .build();
    return s3.putObject(request, path).thenApply(response -> getUri(key)).whenComplete((r, e) -> {
      if (e == null) {
        logger.info("Completed S3 upload of {} in {}", key, stopwatch);
      } else {
        logger.error("Failed to upload to S3: {}", key, e);
      }
    });
  } catch (IOException e) {
    throw new UncheckedIOException(e);
  }
}
```

```java
// Workflow's inner methods

/**
  * Uploads the file asynchronously.
  *
  * @param key the remote storage key
  * @param file the local file to upload
  * @return a handle to the upload
  */
public URI upload(String key, Path file) {
  URI uri = s3.getUri(key);
  files.synchronous().put(uri, file);
  pendingUploads.add(s3.upload(key, file));
  return uri;
}

/** Waits until all of the pending uploads have completed successfully. */
public void await() {
  CompletableFutures.allOf(pendingUploads)
      // .orTimeout(...) to be added to resolve & good practice
      .thenRun(this::invalidate).join();
  pendingUploads.clear();
}

/** Requests an invalidation of the replaced files from the CDN. */
private void invalidate() {
  if (!invalidations.isEmpty()) {
    s3.invalidate(invalidations);
    invalidations.clear();
  }
}
```

```
"Job: entity.0289a7dd-b83b-430a-aca4-7bc8f006ecf9_at_2018-09-28T05:57:43.309776Z" #43 daemon prio=5 os_prio=0 tid=0x00007fce98064800 nid=0x7947 waiting on condition  [0x00007fce735f4000]
   java.lang.Thread.State: WAITING (parking)
	at jdk.internal.misc.Unsafe.park(java.base@10.0.2/Native Method)
	- parking to wait for  <0x0000000655d8fc18> (a java.util.concurrent.CompletableFuture$Signaller)
	at java.util.concurrent.locks.LockSupport.park(java.base@10.0.2/LockSupport.java:194)
	at java.util.concurrent.CompletableFuture$Signaller.block(java.base@10.0.2/CompletableFuture.java:1796)
	at java.util.concurrent.ForkJoinPool.managedBlock(java.base@10.0.2/ForkJoinPool.java:3156)
	at java.util.concurrent.CompletableFuture.waitingGet(java.base@10.0.2/CompletableFuture.java:1823)
	at java.util.concurrent.CompletableFuture.join(java.base@10.0.2/CompletableFuture.java:2043)
	at ...pipeline.context.Storage.await(Storage.java:174)
	at ...pipeline.Pipeline.lambda$run$0(Pipeline.java:98)
	at ...pipeline.Pipeline$$Lambda$1352/1423995213.run(Unknown Source)
```

**Comment by dagnir** [comment]
@ben-manes Did you experience the above behavior in `preview-11` as well? Or just `preview-12`?

**Comment by ben-manes** [comment]
I'm not sure. I am now on `preview-11` as it is more stable and has the `PUT` and download issues. Unfortunately `preview-12` was too unstable that I had to rollback, so I don't know if it fixed the `PUT` problem but exacerbated the `GET` ones. Thankfully these did help identify where I could add better robustness so it wasn't all bad. When I do upgrade again, hopefully even with some instabilities it won't require manually resolving.

**Comment by dagnir** [comment]
Gotcha. Thank you for the thorough report! We'll continue to investigate this.

**Comment by ben-manes** [comment]
I worked around many of the issues via retrials, caching, careful file handling, and using sdk v2.0-11. One issue that does occur still is corrupt downloads, causing a retrial to be needed on a different machine on the cluster. That mostly happens automatically, with error reporting to capture the exceptions. It appears that the files may only be partially downloaded by the time the future completes, as in a race on the consumer and flushing the file to disk.

```
Failed to read /tmp/cache_14379695691612351220/8c70fbd2-d09f-4792-b706-85b36e2ee239/dc80554b-65b2-47d2-9cec-58d729dda8e8/component/file/1541788492940/7afc36c4-b088-40b6-b1cd-b2fa417ee6e0/image.webp (243271 bytes)
```

The file size on S3 is 249,542 bytes, meaning that 6271 bytes are missing. That makes sense since Java's buffered I/O is 8kb, so the last flush raced with the business logic. I think that I can detect and retry by checking the response's contentLength with the actual byte size, and add a delay before retrying if the sync doesn't occur.

I also think that on a download failure, the async client propagates the error before deleting the destination file. If using the same destination in a retry loop, this can result in a success followed by the deletion, so the file disappears. I resolved that by always writing to a temporary destination each time, and then moving the file to the expected output when successful.

The current version of the S3 code client is:

```java

/**
 * Creates a file on S3 with the given key as the file name. The request may fail if the key
 * exists and the bucket does not allow overwrites (it is advisable to include a UUID as part of
 * the key). Note that the URI may be optimistically obtained using {@link #getUri(String)}.
 *
 * @param key the name of the file in the bucket
 * @param path the path to the file
 * @return the location to the file
 */
public CompletableFuture<PutObjectResponse> upload(String key, Path path) {
  try {
    var stopwatch = Stopwatch.createStarted();
    var request = PutObjectRequest.builder()
        .cacheControl("max-age=" + cacheControl().toSeconds())
        .contentType(tika().detect(path))
        .key(normalizeKey(key))
        .acl(PUBLIC_READ)
        .bucket(bucket())
        .build();
    return s3().putObject(request, path).whenComplete((r, e) -> {
      if (e == null) {
        logger.info("Completed S3 upload of {} in {}", key, stopwatch);
      } else {
        logger.error("Failed to upload to S3: {}", key, e);
      }
    });
  } catch (IOException e) {
    throw new UncheckedIOException(e);
  }
}

/**
 * Creates a file on locally with the contents of the S3 object for the given key.
 *
 * @param directory the download directory
 * @param key the name of the file in the bucket
 * @return the location to the file
 */
public Path download(Path directory, String key) {
  var request = GetObjectRequest.builder().bucket(bucket()).key(normalizeKey(key)).build();
  Stopwatch stopwatch = Stopwatch.createStarted();
  try {
    Path destination = directory.resolve(request.key());
    Files.createDirectories(destination.getParent());
    var retryPolicy = new RetryPolicy()
        .withDelay(1, TimeUnit.SECONDS)
        .withMaxRetries(3);
    Failsafe.with(retryPolicy).run(() -> {
      Path path = destination.resolveSibling(UUID.randomUUID() + ".tmp");
      s3().getObject(request, path).join();

      // https://github.com/aws/aws-sdk-java-v2/issues/721
      checkState(Files.exists(path) && (Files.size(path) > 0L), "File not found: %s", path);
      Files.move(path, destination, REPLACE_EXISTING);
    });
    logger.info("Completed S3 download of {} in {}", request.key(), stopwatch);
    return destination;
  } catch (FailsafeException | IOException e) {
    throw new RuntimeException("Failed to download from S3: " + request.key(), e);
  }
}
```

**Comment by varunnvs92** [comment]
I believe both issues have been fixed in the latest release `2.0.0-preview-13`. I was able to repo the incomplete file download issue on preview-11 but not on preview-13. We had multiple refactoring around the async code which likely fixed both issues.

Can you try with the latest version and verify the behavior. If you are still seeing the issues, please provide a repo code. Thank you for the opening the issue with detailed reports. I will leave the ticket open awaiting your verification.

**Comment by ben-manes** [comment]
Actually, I'll reopen just to be safe. I don't think this code looks right to me still.

https://github.com/aws/aws-sdk-java-v2/blob/b1d0e17c37d31e4cc07c2e4035d9972405c74da6/core/sdk-core/src/main/java/software/amazon/awssdk/core/internal/async/FileAsyncResponseTransformer.java#L157-L167

Doesn't this mean that the future may be completed before the last write completes, resulting in the incomplete file being seen by the client on the future callback? I would expect that `closeOnLastWrite` would perform the `future.complete(null)` in that condition. Obviously I don't know this code base, though.

**Comment by shorea** [comment]
@ben-manes  that does look like a problem, we should probably complete the future in the close method.

**Comment by ben-manes** [comment]
In v13, the errors are occurring but being handled by the retrial logic. I added `onFailedAttempt` callback to log.

1. For the race condition (thanks for the quick PR to fix this):
```
Incomplete download for key /87103dd8-809c-4071-908d-31607a9c53ad/component/image/1542320492993/93923f14-6695-4364-9acb-9f71a7a4cc27/image.txt
```

2. `NoSuchFileException` is being thrown by `FileAsyncResponseTransformer#exceptionOccurred`. This should use `Files.deleteIfExists` as perhaps it can be called multiple times?

3. `ReadTimeoutException`, `ConnectTimeoutException`, and `SSLException: SSLEngine closed already` also appear. But they are from the sdk, retried, logged. I don't think they bubble to the caller so expected in a chaotic DC environment.

**Comment by varunnvs92** [comment]
>> This should use Files.deleteIfExists as perhaps it can be called multiple times?
Makes sense, it is safer to use `Files.deleteIfExists`. I will update the PR

**Comment by ben-manes** [comment]
Thanks for the quick responses.

There isn't a snapshot published (e.g. to Sonatype) and I'm having trouble getting jitpack to resolve correctly. So I'll reopen as needed.

**Comment by ben-manes** [comment]
I tried updating to 2.1.0 and locally incomplete download problem appears worse. Previously it would always resolve to the full download after the waiting. Now it is stuck as incomplete, as if the last write seems to never be flushed. The only change was updating the dependency locally. Unfortunately our fixes seems to have made this a blocking issue from what I can tell so far.

```
java.lang.IllegalStateException: Expected 520348 bytes but was 520332: /var/folders/f1/hl_nwg_13s71wfzqf6gwnknr0000gn/T/cache_14504199027799570231/f5075138-ab7e-4d9a-a471-1e16e5ebf9e6/b0f08d3b-49d2-4a1c-a5bc-ab44e63efc43/component/file/1542693894515/d62c76ae-6ef6-45f6-a39a-0527b4c28703/fbc2e98a-07d2-499c-b99b-517f36cbcc93.tmp
```

**Comment by varunnvs92** [comment]
Is this happening occasionally or for each download? Is the test code same as in your earlier comment? If not, Can you share updated test code to repo on our end?

**Comment by spfink** [comment]
The 16 bytes is the Md5 checksum of the object. We are stripping those bytes off the response but it appears S3 may not be properly setting the content length on the response. Should be able to get this fixed easily.

**Comment by spfink** [comment]
https://github.com/aws/aws-sdk-java-v2/pull/855

**Comment by ben-manes** [comment]
Thanks! I'll watch #855 and try again whenever its been released. I think if you setup deploying a `SNAPSHOT` that would help, or else I'll wait for releases to verify.

I did see hanging up on uploads with `2.1.0`, but I haven't debugged that yet. Not sure if it was a local problem or a regression. So just fyi.
